File: leetcode/arrays/49-anagrams.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string, vector<string>> um;
        vector<vector<string>> result;

        for(string &s: strs){
            string temp = s;
            sort(temp.begin(), temp.end());

            um[temp].push_back(s);
        }

        for(auto &s: um){
            string key = s.first;
            vector<string> children = s.second;
            
            result.push_back(children);
        }

        return result;
    }
};

File: leetcode/arrays/49-anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> um = new HashMap<>();
        List<List<String>> result = new ArrayList<>();

        for (String str : strs) {
            char[] schar = str.toCharArray();
            Arrays.sort(schar);

            String key = String.valueOf(schar);

            if (!um.containsKey(key)) {
                um.put(key, new ArrayList<>());
            }
            um.get(key).add(str);
        }

        for (List<String> s : um.values()) {
            result.add(s);
        }
        return result;
    }
}